src/util/psbt/map/global.rs

```diff
@@ -113,6 +155,54 @@ impl Map for Global {
             });
         }
 
+        // BIP 174: The Combiner must remove any duplicate key-value pairs, in accordance with
+        //          the specification. It can pick arbitrarily when conflicts occur.
+
+        // Keeping the highest version
+        self.version = cmp::max(self.version, other.version);
+
+        // Merging xpubs
+        for (xpub, (fingerprint1, derivation1)) in other.xpub {
+            match self.xpub.entry(xpub) {
+                Entry::Vacant(entry) => {
+                    entry.insert((fingerprint1, derivation1));
+                },
+                Entry::Occupied(mut entry) => {
+                    // Here in case of the conflict we select the version with algorithm:
+                    // 1) if fingerprints are equal but derivations are not, we report an error; otherwise
+                    // 2) choose longest unhardened derivation; if equal
+                    // 3) choose longest derivation; if equal
+                    // 4) choose largest fingerprint using binary ordering; if equal
+                    // 5) do nothing
+                    let (fingerprint2, len2, normal_len2) = {
+                        // weird scope needed for rustc 1.29 borrow checker
+                        let (fp, deriv) = entry.get().clone();
+                        (fp, deriv.len(), deriv.into_iter().rposition(ChildNumber::is_normal))
+                    };
+                    let len1 = derivation1.len();
+                    let normal_len1 = derivation1.into_iter().rposition(ChildNumber::is_normal);
+                    match (normal_len1.cmp(&normal_len2), len1.cmp(&len2), fingerprint1.cmp(&fingerprint2)) {
+                        (Ordering::Equal, Ordering::Equal, Ordering::Equal) => {},
```

Comment:
Their lengths are equal and their fingerprints are equal. How else could they differ?

Reply:
Ah, as you observe below it's pretty easy to collide fingerprints. That's fair.

Reply:
Not only that. But you also just compare derivation path length. Assuming the same fingerprint/key the both derivation paths `m/0'/0` and `m/1'/2` denote different keys but have the same `normal_len` and `len`, which is all that's being tested for if I'm not mistaken.

Reply:
Fingerprints can't collide here, since if we also have xpub equality. Basically I will add comparison between full derivation paths, and if they are equal it will mean that we have xpubs equal, derivation paths equal and equal fingerprints will mean equality, it can't be a collision anymore.

Reply:
Oh, right, I didn't take into account that the xpub has to be equal in the first place.

Reply:
Or wait, that assumes that the fingerprint+derivation path has to actually lead to the same pubkey, but what about attacker-provided PSBTs where that isn't the case? Can anything bad happen in that case?

Reply:
Fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/499/commits/21c11e3315d6753b868abc7a61a0772c46f5646c